### PR TITLE
Replace macro for unit tests with a new function

### PR DESCRIPTION
**Pull request by Luni-4**
This PR replaces the macro for unit tests with a new function. These changes have been made for the following reasons:
- The old macro was hard to understand and complicated, moreover it was prone to errors
- The new function allows to customize tests, changing the code inside the anonymous function passed as argument
- Snapshots, created through the `insta` crate, allow to better visualize differences as plain text in case of errors during tests execution, in addition the comparison is made with the `Serializable` representation of metrics, so we can also check if serialized data is correct using the same unit test.

No optimization loss has been found during the replacement

### CODE REVIEW

File: src/metrics/exit.rs [hunk 1]
```diff
@@ -36,7 +36,7 @@ impl Serialize for Stats {
         S: Serializer,
     {
         let mut st = serializer.serialize_struct("nexits", 4)?;
-        st.serialize_field("sum", &self.exit())?;
+        st.serialize_field("sum", &self.exit_sum())?;
```
Comment (marco-c) [inline src/metrics/exit.rs:39]: this change is out of scope for this PR

Comment (Luni-4) [reply src/metrics/exit.rs:39]: I was able to find that error through these new tests, so it's not completely out of scope. What's the best way to proceed here? If I change that line, tests will surely fail

Comment (Luni-4) [reply src/metrics/exit.rs:39]: The problem is that old macro tests act on functions, they do not consider the serialization output, so if we want to be precise we need to: 

1. Remove that line
2. Merge this PR even if tests are failing
3. Fix tests in another PR

Comment (marco-c) [reply src/metrics/exit.rs:39]: What happens exactly if you just change `st.serialize_field("sum", &self.exit())?;' to `st.serialize_field("sum", &self.exit_sum())?;`? Can't we do that and fix the tests, then apply the changes you are proposing in this PR?

Comment (Luni-4) [reply src/metrics/exit.rs:39]:
```console
failures:

---- metrics::exit::tests::python_simple_function stdout ----
Stats { exit: 0, exit_sum: 1, total_space_functions: 1, exit_min: 0, exit_max: 1 }
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Summary ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot: python_simple_function
Source: src/metrics/exit.rs:272
────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Expression: metric.nexits
────────────────────────────────────────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬───────────────────────────────────────────────────────────────────────────────────────────────────
    0     0 │ {
    1       │-  "sum": 1.0,
          1 │+  "sum": 0.0,
    2     2 │   "average": 1.0,
    3     3 │   "min": 0.0,
    4     4 │   "max": 1.0
    5     5 │ }
────────────┴───────────────────────────────────────────────────────────────────────────────────────────────────
To update snapshots run `cargo insta review`
Stopped on the first failure. Run `cargo insta test` to run all snapshots.
thread 'metrics::exit::tests::python_simple_function' panicked at 'snapshot assertion for 'python_simple_function' failed in line 272', /home/.cargo/registry/src/github.com-1ecc6299db9ec823/insta-1.29.0/src/runtime.rs:569:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- metrics::exit::tests::python_more_functions stdout ----
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Summary ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot: python_more_functions
Source: src/metrics/exit.rs:298
────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Expression: metric.nexits
────────────────────────────────────────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬───────────────────────────────────────────────────────────────────────────────────────────────────
    0     0 │ {
    1       │-  "sum": 2.0,
          1 │+  "sum": 0.0,
    2     2 │   "average": 1.0,
    3     3 │   "min": 0.0,
    4     4 │   "max": 1.0
    5     5 │ }
────────────┴───────────────────────────────────────────────────────────────────────────────────────────────────
To update snapshots run `cargo insta review`
Stopped on the first failure. Run `cargo insta test` to run all snapshots.
thread 'metrics::exit::tests::python_more_functions' panicked at 'snapshot assertion for 'python_more_functions' failed in line 298', /home/.cargo/registry/src/github.com-1ecc6299db9ec823/insta-1.29.0/src/runtime.rs:569:9

---- metrics::exit::tests::python_nested_functions stdout ----
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Summary ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot: python_nested_functions
Source: src/metrics/exit.rs:324
────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Expression: metric.nexits
────────────────────────────────────────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬───────────────────────────────────────────────────────────────────────────────────────────────────
    0     0 │ {
    1       │-  "sum": 2.0,
          1 │+  "sum": 0.0,
    2     2 │   "average": 0.5,
    3     3 │   "min": 0.0,
    4     4 │   "max": 1.0
    5     5 │ }
────────────┴───────────────────────────────────────────────────────────────────────────────────────────────────
To update snapshots run `cargo insta review`
Stopped on the first failure. Run `cargo insta test` to run all snapshots.
thread 'metrics::exit::tests::python_nested_functions' panicked at 'snapshot assertion for 'python_nested_functions' failed in line 324', /home/.cargo/registry/src/github.com-1ecc6299db9ec823/insta-1.29.0/src/runtime.rs:569:9

---- metrics::exit::tests::java_simple_function stdout ----
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Summary ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot: java_simple_function
Source: src/metrics/exit.rs:366
────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Expression: metric.nexits
────────────────────────────────────────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬───────────────────────────────────────────────────────────────────────────────────────────────────
    0     0 │ {
    1       │-  "sum": 1.0,
          1 │+  "sum": 0.0,
    2     2 │   "average": 1.0,
    3     3 │   "min": 0.0,
    4     4 │   "max": 1.0
    5     5 │ }
────────────┴───────────────────────────────────────────────────────────────────────────────────────────────────
To update snapshots run `cargo insta review`
Stopped on the first failure. Run `cargo insta test` to run all snapshots.
thread 'metrics::exit::tests::java_simple_function' panicked at 'snapshot assertion for 'java_simple_function' failed in line 366', /home/.cargo/registry/src/github.com-1ecc6299db9ec823/insta-1.29.0/src/runtime.rs:569:9

---- metrics::exit::tests::java_split_function stdout ----
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Summary ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot: java_split_function
Source: src/metrics/exit.rs:394
────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Expression: metric.nexits
────────────────────────────────────────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬───────────────────────────────────────────────────────────────────────────────────────────────────
    0     0 │ {
    1       │-  "sum": 2.0,
          1 │+  "sum": 0.0,
    2     2 │   "average": 2.0,
    3     3 │   "min": 0.0,
    4     4 │   "max": 2.0
    5     5 │ }
────────────┴───────────────────────────────────────────────────────────────────────────────────────────────────
To update snapshots run `cargo insta review`
Stopped on the first failure. Run `cargo insta test` to run all snapshots.
thread 'metrics::exit::tests::java_split_function' panicked at 'snapshot assertion for 'java_split_function' failed in line 394', /home/.cargo/registry/src/github.com-1ecc6299db9ec823/insta-1.29.0/src/runtime.rs:569:9


failures:
    metrics::exit::tests::java_simple_function
    metrics::exit::tests::java_split_function
    metrics::exit::tests::python_more_functions
    metrics::exit::tests::python_nested_functions
    metrics::exit::tests::python_simple_function

test result: FAILED. 237 passed; 5 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.0s
```

**Review by marco-c — DISMISSED** [review]

**Review by marco-c — APPROVED** [review]
